cli-plugins/manager/manager.go

```diff
@@ -41,6 +43,79 @@ func getPluginDirs(dockerCli command.Cli) []string {
 	return pluginDirs
 }
 
+func addPluginCandidatesFromDir(res map[string][]string, d string) error {
+	dentries, err := ioutil.ReadDir(d)
+	if err != nil {
+		return err
+	}
+	for _, dentry := range dentries {
+		switch dentry.Mode() & os.ModeType {
+		case 0, os.ModeSymlink:
+			// Regular file or symlink, keep going
+		default:
+			// Something else, ignore.
+			continue
+		}
+		name := dentry.Name()
+		if !strings.HasPrefix(name, NamePrefix) {
+			continue
+		}
+		name = strings.TrimPrefix(name, NamePrefix)
+		if runtime.GOOS == "windows" {
+			exe := ".exe"
+			if !strings.HasSuffix(name, exe) {
+				continue
+			}
+			name = strings.TrimSuffix(name, exe)
+		}
+		res[name] = append(res[name], filepath.Join(d, dentry.Name()))
+	}
+	return nil
+}
+
+// listPluginCandidates returns a map from plugin name to the list of (unvalidated) Candidates. The list is in descending order of priority.
+func listPluginCandidates(dirs []string) (map[string][]string, error) {
+	result := make(map[string][]string)
+	for _, d := range dirs {
+		// Silently ignore any non-directories in the set of paths
+		if fi, err := os.Stat(d); err == nil && !fi.IsDir() {
```

Comment:
Wondering if we should ignore errors due to permissions as well (iow; if a user does not have permissions on a directory, just skip that directory, and the plugins)

Reply:
That's coming close to skipping on any `Stat` error at all, maybe that's the right thing.

Reply:
I went with skipping anything which gave an error from `os.Stat` or wasn't a dir.